### PR TITLE
[#55] Update to mxnet v0.9.1

### PR DESCRIPTION
**Pull request by conopt**
1. Port optimizer to mxnet.cpp
2. Add imperative operator support
3. Fix op generator

Tested using lenet_with_mxdataiter

Bugs from upstream:
1. ~nnvm doesn't parse empty tuple correctly (affets op initialization when the default values of shape parameters are empty)~ (FIXED)
2. missing argument information for `Activation` (affects op generator only)
3. missing type name for `nnvm::Tuple<int>`, which is used in operator `Reshape` as type of `target_shape` (ditto)

### CODE REVIEW

**Comment by mz24cn** [comment]
So, have you finished your work and now wait upstream fix their bugs?

My program failed in include/mxnet-cpp/operator.hpp, line 22:
handle_ = op_map_->GetSymbolCreator(operator_name);
operator_name is "**_Plus**", while **handle_** gets a null.

As for my program, LSTM requires a plus operation ("auto gates = i2h + h2h;", i2h and h2h are two FullyConnected Symbol), it throws Exception due to reason above.

Could you please investigate it or give some hints if I should modify code? The code runs properly for the mxnet version before NNVM refactored.
thank you.

**Comment by mz24cn** [comment]
I have solved the issue above. include\mxnet-cpp\op_suppl.h should be improved since operator names are refactored. For example, line 22,
   return Operator("_Plus")(lhs, rhs)
need to be changed to:
   return Operator("elemwise_add")(lhs, rhs)
All title case names seem to be lower case in recent version.

**Comment by conopt** [comment]
@mz24cn "_Plus" still exists, it's the api to get operator changes, this pull request has included such change. By the way, can you checkout my branch and try your code again? If it goes well, I will merge this PR.

**Comment by mz24cn** [comment]
No, I already changed to your repository (lx75249/v0.9.1) and fetched your code up to date. The Optimizer issue is gone. 
Then the program breaks on "auto gates = i2h + h2h;". I modified op_suppl.h descibed as above, the operator name issue is gone. 
Now the program still breaks on "label = transpose(label);"

**Comment by conopt** [comment]
@mz24cn but line 22 in operator.hpp in my branch doesn't match your description, I think you have the wrong code base?

**Comment by conopt** [comment]
@mz24cn The transpose bug is caused by nnvm, which was just fixed this morning. You need to update and rebuild nnvm and mxnet locally.

**Comment by mz24cn** [comment]
My IDE eclipse doesn't find your today's code. I will use git bash to fetch again and recompile mxnet and my code. I will report you the result tonight.
thank you.

**Comment by mz24cn** [comment]
Hi, I rebased my code to your current v0.9.1. There are still issues.
1, Optimizer and **_Plus** operator name issues are okay, passed.
2, "label = transpose(label);" breaks. Changed it to be "label = transpose(label, Shape(sequence_length, batch_size));", passed. I think it is not a proper behaviour since transposing a Symbol or NDArray is clear and straightforward.
3, “label = Reshape(label, Shape(batch_size * sequence_length));” breaks. Changed it to be "label = Reshape(label, Shape(0, 0), false, Shape(batch_size * sequence_length));", passed. I think it is not a good parameter list since target_shape and keep_highest are deprecated, why not make they stay at tail and keep a default value.
4, Now the program breaks in MXSymbolInferShape (from RNN.SimpleBind to Symbol::InferExecutorArrays to it). I cannot see details since debug version libmxnet.dll still in compiling. I guess it concerns with transpose or Reshape.

**Comment by conopt** [comment]
@mz24cn you need to update the nnvm submodule of mxnet, otherwise almost all default shape parameters are buggy, like in 2 and 3.

**Comment by mz24cn** [comment]
And for include\mxnet-cpp\operator.hpp, I add one line "**#include <iterator>"** otherwise back_inserter cannot pass the compilation uder VS2015.

**Comment by mz24cn** [comment]
#include &lt;iterator&gt;

**Comment by mz24cn** [comment]
Recompile all but everything the same.
You can use five lines to verify it:
void main(int argc, char** argv)
{
	auto label = Symbol::Variable("softmax_label");
	label = transpose(label);
}

**Comment by conopt** [comment]
@mz24cn Compare your mxnet/nnvm/include/nnvm/tuple.h with https://github.com/dmlc/nnvm/blob/master/include/nnvm/tuple.h#L236. I think you used outdated nnvm.

**Comment by mz24cn** [comment]
I got it. I should manually update nnvm from dmlc/nnvm, not from dmlc/mxnet/nnvm. I will try.

**Comment by mz24cn** [comment]
Now the program breaks on "exe->UpdateAll(opt, learning_rate, weight_decay);". Line 113, include/mxnet-cpp/optimizer.hpp, states_ contains one element: NULL. 

As for transpose and Reshape I use code as follows:
	label = transpose(label);
	label = Reshape(label, Shape(0));
and passed.

**Comment by conopt** [comment]
@mz24cn where can I see your code?

**Comment by mz24cn** [comment]
https://github.com/mz24cn/MXNet.cpp/blob/master/example/charRNN.cpp 
If you need debug in windows, in windows/vs/charRNN has a project.
The repository is not based on your fork, but actually I am not working on this repository, but on yours v0.9.1.

**Comment by conopt** [comment]
@mz24cn it runs well on linux, outputing epoch 1 Train-Perplexity=1895.42. I will try on windows later days.
